### ISSUE TITLE
Random upload failures

### ISSUE THREAD

**Issue by scossu** [aic-collections/aicdams-lakeshore#189]
In file upload page, when multiple files are selected, some (up to 50% of them) fail to upload. 

The same files may upload with no error if the same operation is repeated, while others that previously uploaded may fail. 

Also noticed that if the upload page is kept open after the failed uploads and the same files are re-added and uploaded, they never fail as far as has been tested.

**Comment by awead** [comment]
@scossu I'll need more information here, such as number of files (exact), types, sizes, etc. If you can provide me with the same files that you're using to test, that would be ideal.

**Comment by scossu** [comment]
@awead I shared the sample files on GDoc. 
I tried different subsets of these, using both single/multi file and folder selections. The error appears with as little as 2 files selected. 

I have a feeling that it is not an issue with the files themselves, but a client/server problem, maybe related to asynchronous operations. 

N.B. I have tested with Firefox only, will test on Chrome soon and report back.

**Comment by awead** [comment]
@scossu one thing that occurred to me is that when you select the asset type, still image vs. text, it applies to all the files uploaded. So if you mix the two types in a single batch, the ones that aren't text or images will fail.

**Comment by scossu** [comment]
All the files I uploaded are jpegs and PDFs which are supposed to be valid image types. Also, even just two jpeg files can have at least one fail. 

Moreover, if I re-select the faiiled files and re-submit after I receive the errors (without navigating out of the upload page or reloading it) all files are uploaded successfully. That is why I am suspecting an AJAX/concurrency issue.

**Comment by scossu** [comment]
A couple more test results:
- Confirmed same behavior on Chrome, in production. 
- Same behavior with text files in production.
- All images fail in test. 

N.B. When a file fails, the file name is not shown so it is hard to tell which needs to be re-submitted. This will be addressed in a separate ticket.

**Comment by awead** [comment]
@scossu I was able to upload all the files in the sample provided in one batch, so it's probably an issue with your server deployment or configuration. I'll need to do some testing on your systems to see what's going on. Any ideas? The batch upload is executing a series of POST requests that upload each file. I'll need to run the same upload process and watch the log files to see what error is being reported.

**Comment by scossu** [comment]
@awead This may likely be an environment issue as you are guessing, especially since the behaviour is not consistent and it differs between production and test in general.

**Comment by awead** [comment]
I'm unable to replicate this problem in production, however, if I upload files in test, I'm seeing this error:

```
E, [2016-01-06T21:18:20.841166 #10612] ERROR -- : GenericFilesController::create rescued Ldp::HttpError
    STATUS: 500 javax.jcr.nodetype.ConstraintViolationException: Cannot find a definition for a child named '{}01' on the node at '/prod1/SI/00' with primary type 'nt:folder' and mixin types: [fedora:Pairtree]
    at org.modeshape.jcr.AbstractJcrNode.nodeDefinition(AbstractJcrNode.java:2911)
    at org.modeshape.jcr.AbstractJcrNode.getDefinition(AbstractJcrNode.java:2835)
    at org.modeshape.jcr.AbstractJcrNode.isCheckedOut(AbstractJcrNode.java:3222)
    at org.modeshape.jcr.AbstractJcrNode.addChildNode(AbstractJcrNode.java:1131)
    at org.modeshape.jcr.AbstractJcrNode.addNode(AbstractJcrNode.java:1072)
    at org.modeshape.jcr.AbstractJcrNode.addNode(AbstractJcrNode.java:986)
    at org.modeshape.jcr.AbstractJcrNode.addNode(AbstractJcrNode.java:101)
    at org.modeshape.jcr.api.JcrTools.findOrCreateNode(JcrTools.java:437)
    at org.modeshape.jcr.api.JcrTools.findOrCreateNode(JcrTools.java:377)
    at org.fcrepo.kernel.modeshape.services.AbstractService.findOrCreateNode(AbstractService.java:50)
    at org.fcrepo.kernel.modeshape.s...
```

Could this be a problem with your instance of Fedora?

**Comment by scossu** [comment]
That sounds like a different error. I will investigate separately in that regard.

**Comment by scossu** [comment]
I just replicated the error. There are some logs on LAKEshore at 2016-01-06T22:26:22 server time.

**Comment by awead** [comment]
@scossu is this on test or production?

**Comment by scossu** [comment]
Production.

**Comment by awead** [comment]
I'm not seeing any failures to upload files. I am seeing lots of errors to the reindexing API:

```
I, [2016-01-06T22:26:22.397493 #17131]  INFO -- : Started POST "/api/reindex" for 198.40.25.3 at 2016-01-06 22:26:22 -0600
I, [2016-01-06T22:26:22.398768 #17131]  INFO -- : Processing by ReindexController#update as HTML
I, [2016-01-06T22:26:22.398832 #17131]  INFO -- :   Parameters: {"_json"=>["SI-000195/content"], "reindex"=>{"_json"=>["SI-000195/content"]}}
D, [2016-01-06T22:26:22.491847 #17131] DEBUG -- :   Load LDP (43.9ms) https://lakesuperior.artic.edu/fcrepo/rest/prod/SI/00/01/95/SI-000195/content Service: 70064111007920
I, [2016-01-06T22:26:22.492422 #17131]  INFO -- : Completed 500 Internal Server Error in 93ms (ActiveRecord: 0.0ms)
F, [2016-01-06T22:26:22.493613 #17131] FATAL -- : 
NoMethodError (undefined method `new' for nil:NilClass):
  app/controllers/reindex_controller.rb:27:in `deleted_from_solr?'
  app/controllers/reindex_controller.rb:22:in `block in reindex_submitted_ids'
  app/controllers/reindex_controller.rb:21:in `each'
  app/controllers/reindex_controller.rb:21:in `reindex_submitted_ids'
  app/controllers/reindex_controller.rb:6:in `update'
```

How have you configured things to call the API? For one thing, the parameters are incorrect, and the calls seem to be very frequent. Is there a reason for this?

**Comment by scossu** [comment]
I would defer to @acoburn for that, but what seems to me to be happening Is that the Indexer is traversing inside the asset. This creates a lot of errors but it seems like the assets themselves are being indexed correctly. 

@acoburn, can you confirm that?

**Comment by acoburn** [comment]
@awead I assume the `Parameters` value should be just `{"_json"=>["SI-000195/content"]}`. In terms of the frequency with which they are called, I can aggregate similar requests so that, e.g. if the same object is updated multiple times in N seconds, only a single reindex call is made. Would that work? And what would be a good value for N? (10 seconds? 5 seconds?). This would introduce more of a lag into the reindexing process, but it also makes things more resource efficient overall.

**Comment by awead** [comment]
The parameters should be `{"_json"=>["SI-000195"]}` without the "content" but I'm wondering if your calls to the reindex API are being triggered by the upload process.

**Comment by scossu** [comment]
@awead, 

> I'm wondering if your calls to the reindex API are being triggered by the upload process. 

That is a very good point. Geneva might be reindexing the assets that have just been indexed by LAKEshore. 

An improvement to the indexer such as https://github.com/fcrepo4-exts/fcrepo-camel-toolbox/issues/67 should take care of both issues. What we want Geneva to trigger reindexing on are only resources created externally to LAKEshore. And those resources can be filtered by RDF type.

**Comment by acoburn** [comment]
It sounds like we should just filter out the `/content` nodes (based on the value of a property in e.g. `rdf:type`) in the solrize route. I can add that later this afternoon (after some meetings).

**Comment by scossu** [comment]
I would actually filter out the assets themselves too, since those are entirely managed and indexed by LAKEshore, no?

If you can implement a configurable whitelist of RDF types for the indexer, the only ones that the Solr indexer needs to process are: 
- `aictype:Actor`
- `aictype:Work`
- `aictype:Exhibition`
- `aictype:Place` (not implemented in Fedora yet)
- `aictype:Shipment`
- `aictype:Transaction`

I could also add an `aicmix:CitiResource` RDF type that would encompass all of the above.

**Comment by awead** [comment]
I have it as `aictype:CitiResrouce`. I didn't see aicmix in the content model. FWIW, if you create a new Work in Hydra, it'll add the intermediary types such as CitiResource and Resource.

**Comment by acoburn** [comment]
@scossu I've added the `rdf:type` filter to the solrizer route using the list of above resource types (not aictype:CitiResource). I didn't make it configurable because it involved a really long XPath statement that is probably best _not_ defined in a config file. However, if it is the case that `aictype:CitiResource` encompasses all of those types, that would make things much cleaner (and I'd make that configurable).

Can you confirm that `aictype:CitiResource` is an existing `rdf:type` on these resources? And can you also confirm that updates are filtering properly?

**Comment by scossu** [comment]
@awead is right, it is currently aictype:CitiResource (although I would like to make it an aicmix later). 

I like this solution too because it is more flexible (for example it allows me to index lists that come from CITI, which have sthe same aictype as LAKEshore lists). 

I will test the updates and let you know if it is working.

**Comment by scossu** [comment]
@acoburn, I can confirm now that Solr reindexing is triggered only for CITI resources now. Thanks. 

I will keep this issue open until I can verify that the root problem (upload errors) is gone.

**Comment by scossu** [comment]
The problem persists with me. The issue stays open but I am pushing it to beta3 in order to gather feedback from other users.

**Comment by scossu** [comment]
Not showing any more after #201.

**Comment by awead** [comment]
Interesting. I'd still love to know what was causing it. Hopefully it won't come back.
